### ISSUE TITLE
Run bcrypt on multiple cores [was: Authentication with Redis fails "due to error"]

### ISSUE THREAD

**Issue by trehn** [vernemq/vernemq#579]
### Environment

- VerneMQ Version: 1.2.3
- OS: Docker

### Configuration

```
  mqtt:
    environment:
      - DOCKER_VERNEMQ_LISTENER__MAX_CONNECTIONS=96000
      - DOCKER_VERNEMQ_LISTENER__NR_OF_ACCEPTORS=100
      - DOCKER_VERNEMQ_MAX_CLIENT_ID_SIZE=32
      - DOCKER_VERNEMQ_MAX_MESSAGE_SIZE=524288
      - DOCKER_VERNEMQ_PERSISTENT_CLIENT_EXPIRATION=1h
      - DOCKER_VERNEMQ_PLUGINS__VMQ_ACL=off
      - DOCKER_VERNEMQ_PLUGINS__VMQ_DIVERSITY=on
      - DOCKER_VERNEMQ_PLUGINS__VMQ_PASSWD=off
      - DOCKER_VERNEMQ_VMQ_DIVERSITY__AUTH_REDIS__ENABLED=on
      - DOCKER_VERNEMQ_VMQ_DIVERSITY__REDIS__HOST=redis
    image: erlio/docker-vernemq:1.2.3
    ports:
      - "8081:8080"
      - "8888:8888"
    restart: always
    ulimits:
      nofile:
        soft: 100000
        hard: 100000
```

### Problem

I have a single client that – after working fine for weeks – can no longer connect. Every attempt results in a CONNACK 5 (unauthorized). I triple-checked the password and manually validated the bcrypt hash in redis.

The only log message I get is:

```
[warning] <0.3218.0>@vmq_mqtt_fsm:check_user:544 can't authenticate client {[],<<"XXX">>} due to error
```

Peeking at the code, the word "error" is formatted into that log line, in place of a more informative description of the problem. It originates somewhere deeper down the stack.

Of course, this only occurs in production. All other clients work just fine and I can't reproduce the problem elsewhere. The only thing that is special about this client is that it may have had a higher volumes of messages in general and misbehaved in the past (excessive reconnects, simultaneous/duplicate connections).

With my very limited understanding of Erlang and VerneMQ, I was unable to figure out where that "error" originated. Without going deeper into why this single client acts up, maybe we can first figure out what the actual error is and where it originates? Better error messages are always a good thing :)

P.S.: This may or may not be related, but I have seen this error sporadically for other clients too:

```
[error] <0.725.0>@vmq_diversity_script:call_function:64 can't call into Lua sandbox for function auth_on_register due to {timeout,{gen_server,call,[<0.601.0>,{call_function,auth_on_register,[{addr,<<"172.18.0.1">>},{port,58256},{mountpoint,<<>>},{client_id,<<"XXX">>},{username,<<"XXX">>},{password,<<"XXX">>},{clean_session,true}]}]}}
```

That doesn't appear to be a persistent problem however.

**Comment by ioolkos** [comment]
@trehn just as a first reaction: thanks for that detailled report.
I'll have to ping @larshesel or @dergraf to check if those 2 errors are related.

**Comment by trehn** [comment]
Just another detail: using Redis' `monitor` command, I can see that VerneMQ is successfully retrieving the correct key for the client.

**Comment by trehn** [comment]
Did some further testing and I have new mysteries to share. I wrote a script that opens 10 connections with 10 unique but static Client IDs (in my setup, Client ID and username are always the same – I don't want a client to be able to successfully connect multiple times). After connecting, each of these 10 clients publishes a message (but that's probably irrelevant).

So here's what happens when I run my script:

```
Jan 15 14:03:22 msg=connected username="6150345103f30aead017d8bc46fa1ded"
Jan 15 14:03:22 msg=connected username="b4a6345276df63bc4d09deec89441d15"
Jan 15 14:03:23 msg=connected username="83209f8fbc77e167faf30cafdc985dcb"
Jan 15 14:03:24 msg=connected username="a0990d822ae0218f575a81c113742d73"
Jan 15 14:03:24 msg=connected username="b86f2e9631bd27f495e6bf84a8f7b5e0"
Jan 15 14:03:25 msg=connected username="06e13e4197e59235132af6be8b25f6c4"
Jan 15 14:03:26 msg=connected username="653af7a6a95f5404eb5c54de38a92c14"
Jan 15 14:03:26 msg=connected username="dc308f077b85f0981c4690fcc7a5ff97"
Jan 15 14:03:26 msg="connection rejected: not authorised" username="b29d63b7d5c41c489fa4a0796ded5726"
Jan 15 14:03:26 msg="connection rejected: not authorised" username="2d7c3526093c92ec2defb80d9d6cc732"
Jan 15 14:03:28 msg=connected username="b29d63b7d5c41c489fa4a0796ded5726"
Jan 15 14:03:29 msg=connected username="2d7c3526093c92ec2defb80d9d6cc732"
```

The first 8 connections succeed immediately, connections 9 and 10 initially receive a CONNACK 5, but successfully retry shortly thereafter. This behavior is exactly the same over numerous attempts.

Here's what `console.log` has to say about this:

```
2018-01-15 13:03:26.702 [error] <0.5759.0>@vmq_diversity_script:call_function:64 can't call into Lua sandbox for function auth_on_register due to {timeout,{gen_server,call,[<0.637.0>,{call_function,auth_on_register,[{addr,<<"172.18.0.1">>},{port,49976},{mountpoint,<<>>},{client_id,<<"b29d63b7d5c41c489fa4a0796ded5726">>},{username,<<"b29d63b7d5c41c489fa4a0796ded5726">>},{password,<<"XXX">>},{clean_session,true}]}]}}
2018-01-15 13:03:26.702 [warning] <0.5759.0>@vmq_mqtt_fsm:check_user:544 can't authenticate client {[],<<"b29d63b7d5c41c489fa4a0796ded5726">>} due to error
2018-01-15 13:03:26.712 [error] <0.5762.0>@vmq_diversity_script:call_function:64 can't call into Lua sandbox for function auth_on_register due to {timeout,{gen_server,call,[<0.637.0>,{call_function,auth_on_register,[{addr,<<"172.18.0.1">>},{port,49988},{mountpoint,<<>>},{client_id,<<"2d7c3526093c92ec2defb80d9d6cc732">>},{username,<<"2d7c3526093c92ec2defb80d9d6cc732">>},{password,<<"XXX">>},{clean_session,true}]}]}}
2018-01-15 13:03:26.712 [warning] <0.5762.0>@vmq_mqtt_fsm:check_user:544 can't authenticate client {[],<<"2d7c3526093c92ec2defb80d9d6cc732">>} due to error
```

Because of a general desire to see the world burn I now start the script a second time in parallel – without stopping the previous 10 connections. Expected behavior would be the two sets of connections continuously overruling each other, disconnecting the older connection on every automatic reconnect.

The result looks like this:

```
Jan 15 14:53:25 msg=connected username="653af7a6a95f5404eb5c54de38a92c14"
Jan 15 14:53:26 msg=connected username="b4a6345276df63bc4d09deec89441d15"
Jan 15 14:53:26 msg=connected username="83209f8fbc77e167faf30cafdc985dcb"
Jan 15 14:53:28 msg=connected username="b86f2e9631bd27f495e6bf84a8f7b5e0"
Jan 15 14:53:29 msg=connected username="06e13e4197e59235132af6be8b25f6c4"
Jan 15 14:53:30 msg=connected username="a0990d822ae0218f575a81c113742d73"
Jan 15 14:53:30 msg="connection rejected: not authorised" username="dc308f077b85f0981c4690fcc7a5ff97"
Jan 15 14:53:30 msg="connection rejected: not authorised" username="2d7c3526093c92ec2defb80d9d6cc732"
Jan 15 14:53:30 msg=connected username="b29d63b7d5c41c489fa4a0796ded5726"
Jan 15 14:53:31 msg=connected username="6150345103f30aead017d8bc46fa1ded"
Jan 15 14:53:31 msg="connection rejected: not authorised" username="653af7a6a95f5404eb5c54de38a92c14"
Jan 15 14:53:32 msg="connection rejected: not authorised" username="b4a6345276df63bc4d09deec89441d15"
Jan 15 14:53:33 msg="connection rejected: not authorised" username="83209f8fbc77e167faf30cafdc985dcb"
Jan 15 14:53:36 msg="connection rejected: not authorised" username="dc308f077b85f0981c4690fcc7a5ff97"
Jan 15 14:53:36 msg="connection rejected: not authorised" username="2d7c3526093c92ec2defb80d9d6cc732"
Jan 15 14:53:36 msg="connection rejected: not authorised" username="b86f2e9631bd27f495e6bf84a8f7b5e0"
Jan 15 14:53:38 msg="connection rejected: not authorised" username="06e13e4197e59235132af6be8b25f6c4"
Jan 15 14:53:38 msg="connection rejected: not authorised" username="653af7a6a95f5404eb5c54de38a92c14"
Jan 15 14:53:39 msg="connection rejected: not authorised" username="b4a6345276df63bc4d09deec89441d15"
Jan 15 14:53:39 msg="connection rejected: not authorised" username="a0990d822ae0218f575a81c113742d73"
Jan 15 14:53:40 msg="connection rejected: not authorised" username="83209f8fbc77e167faf30cafdc985dcb"
Jan 15 14:53:40 msg="connection rejected: not authorised" username="b29d63b7d5c41c489fa4a0796ded5726"
Jan 15 14:53:41 msg="connection rejected: not authorised" username="6150345103f30aead017d8bc46fa1ded"
Jan 15 14:53:43 msg="connection rejected: not authorised" username="dc308f077b85f0981c4690fcc7a5ff97"
Jan 15 14:53:43 msg="connection rejected: not authorised" username="2d7c3526093c92ec2defb80d9d6cc732"
Jan 15 14:53:43 msg="connection rejected: not authorised" username="b86f2e9631bd27f495e6bf84a8f7b5e0"
Jan 15 14:53:45 msg="connection rejected: not authorised" username="06e13e4197e59235132af6be8b25f6c4"
Jan 15 14:53:46 msg="connection rejected: not authorised" username="a0990d822ae0218f575a81c113742d73"
Jan 15 14:53:47 msg="connection rejected: not authorised" username="b29d63b7d5c41c489fa4a0796ded5726"
Jan 15 14:53:48 msg="connection rejected: not authorised" username="653af7a6a95f5404eb5c54de38a92c14"
Jan 15 14:53:48 msg="connection rejected: not authorised" username="6150345103f30aead017d8bc46fa1ded"
```

In `console.log` there is a matching number of occurrences of the above error. As you can see, the situation degrades significantly. The rejections now occur in "phases" for both sets of clients: for a couple of seconds, all connections succeed (as expected). But after a while, all connections get rejected for a couple of seconds. This cycle keeps repeating itself.

Now for the really bad part: during the phases with rejections, even unrelated clients (that are not part of my script) get rejected. Did I just DoS the broker with just 20 connections using default reconnect settings from paho libs? Existing connections appear unaffected though.

**Comment by ioolkos** [comment]
you got burning desires :)
No, seriously, thanks for the detailled report. So, the auth_on_register timeouts in vmq_diversity must be the root cause here, I guess. 
Can your share your script, or do you think it would not help us anyway due to your specific plugin?
@larshesel @dergraf a penny for your thoughts! :)

**Comment by dergraf** [comment]
@trehn how do you generate the bcrypt hasches?

**Comment by trehn** [comment]
@dergraf with https://pypi.python.org/pypi/bcrypt/3.1.0

```python
redis.set(
    '["","{0}","{0}"]'.format(username),
    dumps({
        'passhash': bcrypt.hashpw(
            password.encode('utf-8'),
            bcrypt.gensalt(prefix=b"2a"),
        ).decode(),
        'publish_acl': [{'pattern': pattern} for pattern in acl_pub],
        'subscribe_acl': [{'pattern': pattern} for pattern in acl_sub],
    }),
)
```

**Comment by dergraf** [comment]
can you do an experiment and use a lower number of `rounds` in your `gen_salt`. It's possible that the hash generation on VerneMQ takes too long and triggers the timeout.

**Comment by trehn** [comment]
@dergraf Good call! Dropping the rounds from the default of 12 down to the minimum of 4 fixes the problem.

Can you increase the timeout?

**Comment by larshesel** [comment]
Hi @trehn it's not possible to configure the timeout at the moment - it's 5 seconds by default I believe. I agree it would be a nice addition to make it configurable. though.

**Comment by trehn** [comment]
@larshesel I have no idea why authentication would take longer than 5 seconds. Checking a bcrypt hash on the same machine takes a little under 300ms in Python. I'd rather not compromise security by decreasing the number of rounds. But being able to overload the authentication system with less than 10 connections seems... wrong. Is Lua somehow a bottleneck here?

**Comment by larshesel** [comment]
Hi @trehn

It might well be Lua which is the bottleneck here. You can try to
authenticate something (bcrypt only gets triggered there's a match in
Redis) and see what the average time to process an authentication
request is using `vmq-admin script show`. On my system it looks like
this for Redis:

```
vmq-admin script show
+------------------+
|      script      |
+------------------+
|share/lua/init.lua|
+------------------+

+--------------------------+----------------+
|          script          |auth_on_register|
+--------------------------+----------------+
|./share/lua/auth/redis.lua|    488232.0    |
+--------------------------+----------------+
```

So it takes about 0.5 seconds to process an authentication request. If
this leads to timeouts depends on two factors: The arrival rate of
authentication requests and how many requests can be handled per
second.

If a single abstract processor where each request takes X seconds to
process then it can handle 1/X requests per second. In the case of
authenticating users let's assume the `bcrypt` part alone takes 0.5
seconds. This gives us an ideal rate of 2 bcrypt opts/second. Any
arrival rate higher than that will mean requests will queue up and
have to wait to be processed.

If we have a timeout of 5 seconds, we would be able to handle up to 10
bcrypt operations arriving at the same time (5s timeout/0.5 seconds
per bcrypt operation) without encountering a timeout. Anything more
than that would experience a timeout. And over time, even having an
arrival rate slightly larger than what the system is capable of will
at some point lead to timeouts.

Of course the above describes an ideal system (ignoring whatever else
might be going on which could have an effect on the throughput) with
only one queue.

`vmq_diversity` will automatically use 10 processes to handle
authentication requests [1] for the built-in auth methods (mongo,
redis, mysql, postgresql), but this will of course be limited by the
actual number of (real) cores in the system. The number of processes
can be controlled/overridden using the hidden configuration option
`nr_lua_states` [2]. This feature was introduced in VerneMQ 1.1.0 (Lua
script balancing).

I hope, at least, this clarifies a bit how things work. How many
actual cores does VerneMQ have available on your system? Are you using
stock redis diversity scripts (where we use 10 processes) or are you
using something custom built (where the default would be 1 process per
script).


[1] https://github.com/erlio/vernemq/blob/master/apps/vmq_diversity/priv/auth/auth_commons.lua#L67

[2] https://github.com/erlio/vernemq/blob/master/apps/vmq_diversity/priv/vmq_diversity.schema#L20

**Comment by trehn** [comment]
@larshesel Thank you so much for taking the time to write this.

```
+-------------------------------------+-----------------+-----------------+
|               script                | on_client_gone  |auth_on_register |
+-------------------------------------+-----------------+-----------------+
|/usr/share/vernemq/lua/auth/redis.lua|74.33333333333333|379751.3333333333|
+-------------------------------------+-----------------+-----------------+
```

I'm testing on a single-core VM right now. I wrote some Python to benchmark bcrypt performance:

```python
#!/usr/bin/env python3
# This script - surprisingly - performs better on multicore machines.
# Probably because bcrypt is not subject to the GIL.
from concurrent.futures import ThreadPoolExecutor
from time import time

import bcrypt


def work():
    start = time()
    bcrypt.checkpw(
        b"foo",
        b'$2b$12$njj2bVcqdxPMc8qEuzCEaebfW4RlzVKdFyuUsPKxdj16hLua69SSC',
    )
    print("{:.3f}: {:.3f}".format(time(), time() - start))


for workers in (1, 5, 10):
    print("\nWORKERS: {}\n".format(workers))
    start = time()
    with ThreadPoolExecutor(max_workers=workers) as thread_pool:
        for i in range(10):
            thread_pool.submit(work)
    print("TOTAL: {:.3f}".format(time() - start))
```

```
WORKERS: 1

1516619915.728: 0.281
1516619916.008: 0.280
1516619916.290: 0.282
1516619916.571: 0.280
1516619916.855: 0.284
1516619917.137: 0.281
1516619917.420: 0.283
1516619917.704: 0.283
1516619917.988: 0.284
1516619918.273: 0.284
TOTAL: 2.827

WORKERS: 5

1516619919.683: 1.406
1516619919.689: 1.415
1516619919.703: 1.414
1516619919.705: 1.407
1516619919.692: 1.407
1516619921.118: 1.421
1516619921.123: 1.431
1516619921.128: 1.423
1516619921.129: 1.412
1516619921.131: 1.418
TOTAL: 2.858

WORKERS: 10

1516619923.932: 2.800
1516619923.944: 2.807
1516619923.972: 2.815
1516619923.995: 2.826
1516619923.984: 2.843
1516619923.984: 2.835
1516619923.994: 2.833
1516619924.000: 2.806
1516619924.001: 2.822
1516619924.002: 2.797
TOTAL: 2.870
```

Not surprising on a single core of course. The last case with 10 workers is what I would expect: all workers finish at around the same time, sharing CPU time evenly between all of them. 

But if that were the case, shouldn't all my connections time out if I do 10 at once? I'm guessing that what really happens is closer to the example with 1 core. That would explain why my first 8 connections *always* succeed and the last two connections *always* fail. 

Wildly speculating: Lua seems unable to pause the worker/state mid-work and always does a complete bcrypt operation before the scheduler switches over to let the next worker/state do it's thing. Am I reading the source correctly in that Lua is calling back to Erlang to do the bcrypting? Maybe that's why.

I'll do some more testing with various amounts of cores.

**Comment by dergraf** [comment]
well the bcrypt part is done in a NIF, it's definitely possible that this blocks other waiting NIF calls on a single core machine... but I guess it is a matter of scheduler config, @larshesel do you have some insights here?

**Comment by nipunkr** [comment]
I am getting the following error while using vernemq:

 [error] <0.11340.0>@vmq_diversity_script:call_function:64 can't call into Lua sandbox for function auth_on_register due to {shutdown,{gen_server,call,[<0.437.0>,{call_function,auth_on_register,[{addr,<<"4*.*1.**5.**">>},{port,37479},{mountpoint,<<>>},{client_id,<<"xxx">>},{username,<<"xxx">>},{password,<<"xxx">>},{clean_session,false}]}]}}

Once I start the server it is fine with few connection like 4-5 times and thereafter broker stops accepting connections.

**Comment by larshesel** [comment]
@trehn it actually looks like the bcrypt NIF library we use is using a single thread for doing the work, in other words, even if we have 10 concurrent processes executing the lua scripts on 10 cores, then the bcrypt NIF library will serialize everything.. so that's less than optimal. Fortunately it's look like it would be relatively easy to change it so the bcrypt library uses more threads (for instance, one per lua script instance). So I wouldn't expect you to get better results, even with more cores.

**Comment by trehn** [comment]
@larshesel Can confirm: looks just the same with 4 cores. Glad you found the issue. Can you put the fix into the next release? Anything else I can do?

**Comment by larshesel** [comment]
You can test when we've made the fix :) But note, even with 4 cores bcrypt will still be slow (by design) and the theoretical max throughput rate assuming 0.5s/bcrypt op will only be 8 bcrypts ops/sec.

**Comment by nipunkr** [comment]
To resolve this slow bcrypt issue on my server I changed the logic at redis.lua

I am storing MD5 hash of oauth2 access token from a popular online auth service. I do exact match of this hash removing the bcrypt logic at redis.lua

https://github.com/erlio/vernemq/blob/master/apps/vmq_diversity/priv/auth/redis.lua

I changed:
 if res.passhash == bcrypt.hashpw(reg.password, res.passhash) then
to 
if res.passhash == reg.password then

**Comment by trehn** [comment]
@larshesel Any news? Utilizing all available cores would really help.

Additionally it would be nice to have some kind of throttling in front of the auth mechanism so VerneMQ doesn't bother running bcrypt if the client has already failed authentication X times in the last Y seconds. But that's a different issue.

**Comment by larshesel** [comment]
We haven't looked into this yet. I've just put it into the 1.7.0 milestone so we can hopefully get this resolved soon.

**Comment by larshesel** [comment]
> Additionally it would be nice to have some kind of throttling in front of the auth mechanism so VerneMQ doesn't bother running bcrypt if the client has already failed authentication X times in the last Y seconds. But that's a different issue.

I think this could be solved generally by #931

**Comment by larshesel** [comment]
Hey @dergraf I was thinking a bit about this and the bcrypt lib supports two modes: `nif` and `port`. The default and current is `nif`. With the `port` mode a pool of ports is supported, the default size of this is 4. I think the port variant is interesting as it makes it easy to utilize all cores for the bcrypt work and since it's done as ports it should be even safer from a scheduler viewpoint than using the `nif` mode. 

What I'm thinking about is how and when would we configure this? I guess we could add some hidden bcrypt cuttlefish options which defaults to the number of cores for the pool size? on the other hand using all cores for this may be really bad for system performance in general... I think therefore if we do something like this that we could add a hidden option which just sets the pool size to 1 and then those who want something different can change it accordingly.

**Comment by larshesel** [comment]
Was closed by accident through a related issue. Not fixed until #1160 has been merged.

**Comment by trehn** [comment]
Thank you!

**Comment by larshesel** [comment]
better very late than never, I guess - hope it helps!
